### PR TITLE
polar and laplacian diffusion filter module class

### PR DESCRIPTION
**Pull request by WillyChap**
This is a module to do diffusion/filtering operators into the model time step. 
This is done in the spectral harmonics space and will need some tuning of substeps ... 

Right now it works well, but might be a little agressive to do to the Q field. 

See: 
https://journals.sagepub.com/doi/full/10.1177/1094342011410088?casa_token=XU77aTGMkjEAAAAA%3AfVKn2V6INlocoCObdb-tyIByO6uYQ6hbn5cZCCvEusw88fJknGsX1aiK-jm_OuRb1sozzwX89Y2h5A 

for mathematical derivation

### CODE REVIEW

File: credit/pol_lapdiff_filt.py [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+    U = BB2[8,:,:]
+    V = BB2[8,:,:]
+    nlat = U.shape[0]
+    nlon = U.shape[1]
+    BB2_tensor=torch.tensor(BB2).clone().detach().to(device)
+    #instantiae this with the original model:
+    DPF = Diffusion_and_Pole_Filter(nlat=nlat,nlon=nlon,device=device)
+
+    #this fixes the divergent modes ... might eb doing too much to Q
+    BBfix=DPF.diff_lap2d_filt(BB2_tensor)
```
Comment (jsschreck) [inline credit/pol_lapdiff_filt.py:420]: Just to confirm, the output of BBfix is still the same shape as BB1 and BB2 here, right? I clone the tensor initially and do not perform any correction on Z500

File: credit/pol_lapdiff_filt.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+        spectral domain operations.
+    
+        Parameters:
+        - T (Tensor): scalar-component of the velocity or vector field.
+        - ind_pol (list/int): Index/indices specifying poles for the filtering process.
+    
+        Returns:
+        - Tuple of Tensors: The modified T components of the scalar field.
+        """
+        T=T.clone()
```
Comment (jsschreck) [inline credit/pol_lapdiff_filt.py]: I think the first clone is redundant?

File: credit/pol_lapdiff_filt.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+    
+    # Calculate the ramp down using a reversed sigmoid function
+    x_down = torch.linspace(-6, 6, ramp_length)
+    sigmoid_down = 1 / (1 + torch.exp(-x_down))
+    array[ramp_down_start:] = torch.flip(sigmoid_down, dims=(0,))
+    
+    # Return the modified array
+    return array
+
+def find_nearest(array, value):
```
Comment (jsschreck) [inline credit/pol_lapdiff_filt.py:93]: Looks like some extra indentation here?

File: credit/pol_lapdiff_filt.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+    Creates an array of specified length with a sigmoid ramp up and down.
+    
+    Parameters:
+    - array_length: The length of the output array.
+    - ramp_length: The length of the ramp up and down.
+    
+    Returns:
+    - An array of the specified length with the described ramp up and down using a sigmoid function.
+    """
+    import numpy as np
```
Comment (jsschreck) [inline credit/pol_lapdiff_filt.py]: Numpy imported already at the top.

File: credit/pol_lapdiff_filt.py [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+            var_spectrum = np.real(2 * s)  #multiply by two in order to conserve variance
+            A = Ck2/torch.sum(Ck2)
+            A[ind_1d:] = 0.
+            Zlow = torch.clone(Z)
+            Zlow[ind_1d:-ind_1d] = 0.0
+            X_filtered11 = torch.real(torch.fft.ifft(Zlow))
+            D[ii,:]=X_filtered11
+        return D
+
+    if len(D.shape)==3:
```
Comment (jsschreck) [inline credit/pol_lapdiff_filt.py:37]: You could combine the two loops here into one, that might help reduce some of the code duplication (you could also put the stuff you use twice here into its own method).

**Review by jsschreck — APPROVED** [review]

File: credit/pol_lapdiff_filt.py (hunk 1 above)
Comment (WillyChap) [reply credit/pol_lapdiff_filt.py:420]: yeah that is the same shape going in. No correction to Z500 looks to be needed.

File: credit/pol_lapdiff_filt.py [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+        if len(chispec.shape) == 1:
+            nt = 1
+            chispec = torch.reshape(chispec, ((ntrunc+1)*(ntrunc+2)//2,1))
+        else:
+            nt = chispec.shape[1]
+    
+        divspec2 = (self.lap*chispec)
+    
+        if idim == 1:
+            uchi, vchi = self.getuv(torch.stack((torch.zeros([divspec2.shape[0],divspec2.shape[1]]).to(device),divspec2)))
```
Comment (jsschreck) [inline credit/pol_lapdiff_filt.py]: All the "device" calls in this method should be "self.device"

**Pull request merged**